examples/nvram/read.c

```diff
@@ -0,0 +1,178 @@
+    printf("* priv: size in bytes of the private part of the key in NV\n");
+    printf("* pub: size in bytes of the public part of the key in NV\n");
+    printf("* -aes/xor: Use Parameter Encryption\n");
+}
+
+int TPM2_NVRAM_Read_Example(void* userCtx, int argc, char *argv[])
+{
+    int rc;
+    WOLFTPM2_DEV dev;
+    WOLFTPM2_KEYBLOB keyBlob;
+    WOLFTPM2_SESSION tpmSession;
+    WOLFTPM2_HANDLE parent;
+    WOLFTPM2_NV nv;
+    TPM2B_AUTH auth;
+    word32 readSize;
+    int paramEncAlg = TPM_ALG_NULL;
+
+    if (argc > 2) {
+        XMEMSET(&keyBlob, 0, sizeof(keyBlob));
+        keyBlob.pub.size = atoi(argv[ARGC_PUB_SIZE]);
+        keyBlob.priv.size = atoi(argv[ARGC_PRIV_SIZE]);
+
+        if (argc == 4) {
+            if (XSTRNCMP(argv[ARGC_PARAM_ENC], "-aes", 4) == 0) {
+                paramEncAlg = TPM_ALG_CFB;
+            }
+            if (XSTRNCMP(argv[ARGC_PARAM_ENC], "-xor", 4) == 0) {
+                paramEncAlg = TPM_ALG_XOR;
+            }
+            argc--;
+        }
+    }
+    else {
+        usage();
+        return 0;
+    }
+
+    if (keyBlob.priv.size == 0 || keyBlob.pub.size == 0) {
+        printf("Specify size of private and public part of the key\n");
+        usage();
+        return 0;
+    }
+
+    XMEMSET(&tpmSession, 0, sizeof(tpmSession));
+    XMEMSET(&parent, 0, sizeof(parent));
+    XMEMSET(&auth, 0, sizeof(auth));
+
+    rc = wolfTPM2_Init(&dev, TPM2_IoCb, userCtx);
+    if (rc != TPM_RC_SUCCESS) {
+        printf("\nwolfTPM2_Init failed\n");
+        goto exit;
+    }
+
+    if (paramEncAlg != TPM_ALG_NULL) {
+        /* Start TPM session for parameter encryption */
+        rc = wolfTPM2_StartSession(&dev, &tpmSession, NULL, NULL,
+                TPM_SE_HMAC, TPM_ALG_CFB);
+        if (rc != 0) goto exit;
+        printf("TPM2_StartAuthSession: sessionHandle 0x%x\n",
+            (word32)tpmSession.handle.hndl);
+        /* Set TPM session attributes for parameter encryption */
+        rc = wolfTPM2_SetAuthSession(&dev, 1, &tpmSession,
+            (TPMA_SESSION_decrypt | TPMA_SESSION_encrypt | TPMA_SESSION_continueSession));
+        if (rc != 0) goto exit;
+    }
+
+    auth.size = sizeof(gNvAuth)-1;
+    XMEMCPY(auth.buffer, gNvAuth, auth.size);
+
+    /* Prepare auth for NV Index */
+    XMEMSET(&nv, 0, sizeof(nv));
+    nv.handle.hndl = TPM2_DEMO_NVRAM_STORE_INDEX;
+    nv.handle.auth.size = auth.size;
+    XMEMCPY(nv.handle.auth.buffer, auth.buffer, auth.size);
+
+    readSize = keyBlob.pub.size;
+    printf("Trying to read %d bytes of public key part from NV\n", readSize);
+    rc = wolfTPM2_NVReadAuth(&dev, &nv, TPM2_DEMO_NVRAM_STORE_INDEX,
+        (byte*)&keyBlob.pub.publicArea, &readSize, 0);
+    if (rc != 0) goto exit;
+
+    readSize = keyBlob.priv.size;
+    printf("Trying to read %d bytes of private key part from NV\n", readSize);
+    rc = wolfTPM2_NVReadAuth(&dev, &nv, TPM2_DEMO_NVRAM_STORE_INDEX,
+        (byte*)&keyBlob.priv.buffer, &readSize, keyBlob.pub.size);
```

Comment:
After loading the `keyBlob` can you add code to demonstrate loading the key?

Reply:
Good point. I will address that in 143 as well